### PR TITLE
Observables/Measurement rework

### PR DESCRIPTION
**Pull request by ffreyer**
Carsten and I had a short discussion about how measurements should work in `MonteCarlo.jl`. Currently one has to define `prepare_observables` to initialize a set of observables and `measure_observables!` to calculate and save the current value of each observable. The key issue with this system is that adding ore removing measurements between runs requires redefining both functions.

A good measurement system should follow these points

* It should be simple to add and remove measurements between runs.
* The overhead of taking a measurement (i.e. what happens every sweep) should be small. 
* It should be easy to define new measurements.
* It should be possible to share data and/or bundle measurements. For example, if one wants to measure `M = sum(conf)` and `M2 = sum(conf)^2` it should be possible to calculate `sum(conf)` once and use the result for both measurements
* It should be clear which measurement belong to which flavor and model. 
* The type of Observable should be easy to change.

I tried to come with something that satisfies these points. The first commit is a proof of concept. The idea is the following:

A measurement is represented by the `Measurement` type and bundles an observable, some, possibly shared, data and few functions `init!`, `update_data!`, `measure!` and `finish!`. Each function except `update_data!` is a function of `obs, data, mc, model, args...` where `obs` and `data` come from the same measurement object. `update_data!` currently does not take `obs` as an argument. A minimal example would be:

```julia
Magnetization = Measurement(
    Observable(Float64, "Magnetization"),
    measure! = (obs, data, mc, model, E) -> push!(obs, sum(mc.conf))
)
```

To share data:

```julia
M_data = Ref(0.0)
M_data_update(data, mc, model, E) = sum(mc.conf)
M = Measurement(
    Observable(Float64, "Magnetization"),
    data = M_data,
    update_data! = M_data_update,
    measure! = (obs, data, mc, model, E) -> push!(obs, data[])
)
M2 = Measurement(
    Observable(Float64, "Magnetization squared"),
    data = M_data,
    update_data! = M_data_update,
    measure! = (obs, data, mc, model, E) -> push!(obs, data[]^2)
)
```

Multiple measurements would be given to the MC algorithm, probably when calling `run!()`. MC then proceeds to call `initialize!(measurements, ...)` which does some safety checks, calls `measurement.init!` for each measurement and generates a `measure!` function. `measure!(measurements, ...)` is then called on every sweep, first calling `measurement.update_data!` and then calling `measurement.measure!` for each measurement. At the end of run, `finish!(measurements, ...)` calls `measurement.finish!()` for all measurements.

The performance is currently slightly better than the old system (for the Ising model).

Concept TODO:
- `initialize()` should make sure that if multiple Measurement's have the same `update_data!` function, it is only called once per sweep. (This is currently the case for `measurement.init!`)
- Should multiple observables be allowed in the measurement struct? This might be more convenient than sharing data via the data field.
- There should probably be an option to only measure every n-th sweep in the Measurement object.
- `initialize!`, `measure!` and `finish!` should probably assume or at least work for dictionaries.

### CODE REVIEW

**Comment by ffreyer** [comment]
Another idea would be to let the user define a type `<: Measurement`, which carries around whatever data is needed to take a measurement, and functions `init!` (or a constructor), `measure!` (or making the type callable) and a `finalize!` to define a measurement.

The problem I had with this was mainly about sharing data between measurements. We can deal with this the same way as above - measurements that share some computation (in `measure!`) can be combined into one measurement, or both measurement could share data via some reference/array given to both.
The first option pushes for repeated code, i.e. define `X` for when you only want X, `Y` for when you want `Y`, `XY` for both. 
The second option makes the initialization more awkward. Preferably, the user would give a type to the package and which would then initialize it. Having a shared data field would mean that the user has to at least partially initialize the Measurement by passing a reference to that data.

But the concept above has the same issues, I believe. So perhaps this is just a more straight forward way to do it. On a side note, one could also generalize the concept above and pass Functors instead of Functions.

File: src/Measurements.jl [hunk 1]
```diff
@@ -1,108 +1,47 @@
-struct Measurement{DataT, ObsT <: MonteCarloObservable.AbstractObservable}
-    obs::ObsT
-    data::DataT
-    init!::Function
-    update_data!::Function
-    measure!::Function
-    finish!::Function
```
Comment (carstenbauer) [inline src/Measurements.jl]: Note that `Function` is an abstract type and should be avoided as a field type.

Comment (ffreyer) [reply src/Measurements.jl]: The code generation in `init!` actually dereferences all the individual `update_data!` and `measure!` calls. Meaning that if you set the `measure!` field to some function `f`, the generated `measure!` will contain `f`, not `measurement.measure!`.

And that is also why the code generation was there in the first place. Iirc, parameterizing on `Function` did not help.

**Comment by ffreyer** [comment]
Commit 2:

Here the idea is that each concrete measurement is given by a type `<: AbstractMeasurement`. `prepare!`, `measure!` and `finish!` need to be implemented for that type. At runtime, a collection of measurements (currently a `Dict{Symbol, AbstractMeasurement}`) can be given to `MC`/`DQMC`. During the simulation the `prepare!`, `measure!` and `finish!` are called at appropriate time on each measurement.

Some thoughts:
* We should allow measurements to be passed as types rather than objects. With a `measurement_type(mc, model)` we should be able to initialize every measurement.
* There should be a way to measure during thermalization. For example, one may want to measure the ising energy to verify that a system has thermalized, or to decide how much thermalization is necessary.

**Comment by ffreyer** [comment]
TODO:
- [x] something like `remove(mc, key, stage=:ME)`
- [x] unit tests
- [x] integration tests (correct measurement results)
- [x] docstrings, probably
- [ ] documentation (in a different pr?)
- [x] generalize dqmc measurements to flavor level (every DQMC simulation has a boson energy, Greens function, right?)
- [ ] generalize Ising measurements further (i.e. to flavor level)
- [x] maybe generalize measurements to work with both `LightObservable` and `Observable`
- [x] update example
- [x] show number of measurements (or some other relevant info) for types `::MC`, `::DQMC`

**Comment by ffreyer** [comment]
> * [ ]  maybe generalize measurements to work with both `LightObservable` and `Observable`

Done for DQMC, not for the IsingModel. For the Ising case some observables must be of type `Observable` for the computation of specific heat and susceptibility. Imo this would be an obvious use-case for [ErrorPropagator](https://github.com/crstnbr/BinningAnalysis.jl/pull/46) from BinningAnalysis.jl, so I'm inclined to leave the IsingModel alone for now.

Once `ErrorPropagator` is accepted and available in MonteCarloObservable we can split measurements apart. I.e. instead of `IsingEnergyMeasurement` we could have:
- `EnergyMeasurement(mc::MC, model)` measuring only total energy
- `SquaredEnergyMeasurement(mc::MC, model)` measuring only squared energy (though I dont see this being useful anymore)
- `EnergyPerSpinMeasurement(mc::MC, model)` measuring only energy per spin
- `SpecificHeatMeasurement(mc::MC, model)` measuring only total energy and total energy squared with ErrorPropagator and deriving `C_V`

which allows for more modularity (without the redundancy of recording the full timeseries for C_V) and simplifies choosing between `LightObservable` and `Observable`. 

There is also no point in

> * [ ]  generalize Ising measurements further (i.e. to flavor level)

if we rework Ising measurements later

**Comment by carstenbauer** [comment]
What's the status here?

**Comment by ffreyer** [comment]
The last TODO should be accurate. Functionally everything is ready.

**Comment by ffreyer** [comment]
Actually, there are a few more convenience things to do:

- [x] some sort of `save_measurements(dqmc)` (this may require an extension to the interface)
- [x] export `measurements(dqmc)`, etc

**Comment by ffreyer** [comment]
To summarize the current implementation:
* A meaurement is a type `<: AbstractMeasurement`, implementing
  * `measure!`
  * optionally: `save!`, `prepare!`, `finish!`, `observables`
* You may also implement `default_measurements` for a given model and MonteCarloFlavor (if not implemented simualtions will be constrcuted w/o measurements)
* A measurement can also happen in the thermalization stage (i.e. to verify that a system thermalized)
* There are a bunch of convenience functions:
  * `push!` and `delete!` to add or remove a measurement from a simulation
  * `measurements` and `observables` to retrieve all measurements/observables in a nested dictionary
  * `observables(measurement)` to retrieve all observables of a measurement in a dictionary
  * `save_measurements` and `load_measurements` to, well, save and load them (`save_measurements` calls `save!`, which by default searches for observables and saves those) (I choose not to name the user facing method `save!` because we may want a "save_parameters_and_observables_and_simualtion_state" at some point)
* There is some custom printing for `AbstractMeasurement; `MC` and `DQMC` now also print the number of measurements
* We have tests for
  * type hierarchy of (concrete) implemented measurements
  * `push!`, `unsafe_push!` and `delete!`
  * `observables` and `measurements`
  * `save_measurements` and `load_measurements`
  * Correctness of measured values (though I did not check this against the previous iteration)

Still debatable / possible TODO
* Do we need `prepare!` and `finish!`?
* Once `BinningAnalysis.ErrorPropagator` is available through MonteCarloObservables, we could use that for Ising measurements and get to `O(log(N))` memory requirement.
* Some Ising measurements can probably be generalized to work with any model (w/ MC)
* documentation

**Comment by carstenbauer** [comment]
Should/can we merge this and discuss the remaining points in separate issues? If so, could you fix the conflicts/rebase?

**Comment by ffreyer** [comment]
Yea, let's merge this

File: src/Measurements.jl [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+
+"""
+    ConfigurationMeasurement(mc, model[, rate=1])
+
+Measures configurations of the given Monte Carlo flavour and model. The rate of
+measurements can be reduced with `rate`. (e.g. `rate=10` means 1 measurement per
+10 sweeps)
+"""
+struct ConfigurationMeasurement <: AbstractMeasurement
+    obs::Observable
```
Comment (carstenbauer) [inline src/Measurements.jl]: `Observable` is no concrete type (it's a `UnionAll`) which is generally bad for performance.
Possible fix: make `ConfigurationMeasurement` parametric.

Comment (ffreyer) [reply src/Measurements.jl]: Ah, so that's why I got a `UnionAll` once when testing `push!`. Fixed now

**Comment by carstenbauer** [comment]
We definitely will have to work on the interface (and some other details) a bit more, but let's merge this for now. Thanks!

**Comment by carstenbauer** [comment]
I just noticed I was to quick in merging this. Apparently your latest fix causes some tests to fail. Can you please fix this?

File: src/Measurements.jl [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+    ConfigurationMeasurement(mc, model[, rate=1])
+
+Measures configurations of the given Monte Carlo flavour and model. The rate of
+measurements can be reduced with `rate`. (e.g. `rate=10` means 1 measurement per
+10 sweeps)
+"""
+struct ConfigurationMeasurement{OT <: Observable} <: AbstractMeasurement
+    obs::OT
+    rate::Int64
+    ConfigurationMeasurement(mc, model, rate=1) = new(
```
Comment (carstenbauer) [inline src/Measurements.jl:104]: Create `o = Observable(...)` first and then use `new{typeof(o)}(...)` to fix the issue.

**Comment by carstenbauer** [comment]
Fixed this in #40

**Pull request merged**